### PR TITLE
Use str::replace instead of replace_in_place

### LINKED ISSUE #148: Not sure if replace_all_in_place is safe
We have this function

```rust
fn replace_all_in_place(&mut self, from: &str, to: &str) -> bool {
        let len = from.len();
        if to.len() != len {
            panic!("mismatch length of from and to string");
        }

        let mut found = false;
        while let Some(start) = self.find(from) {
            let end = start + len;

            // NOTE(nlordell): safe since the to string is valid utf-8 and
            //   `str::len()` returns byte length and not character length
            let section = unsafe { self[start..end].as_bytes_mut() };
            section.copy_from_slice(to.as_bytes());

            found = true
        }

        found
    }
```

The comment is correct about the length but I am not sure if replacement operation satisfies the condition that all strings are valid utf8. I don't have a counter example but utf8 is not a trivial format so it could be possible.

It could be enough to assert that both strings are ascii or we could check for valid utf8 again after the replacement. It would be nice if it was possible without unsafe, but not sure.

Or its already correct to someone, that knows utf8 better.

### PR DESCRIPTION
**Pull request by e00E**
In order to avoid unsafe.

fixes #148

### CODE REVIEW

**Comment by e00E** [comment]
Should I move `to_fixed_hex` into `bytecode.rs` too?

File: common/src/bytecode.rs [hunk 1]
```diff
@@ -72,10 +72,10 @@ impl Bytecode {
         //   `LinkedContract` contract for and example of how it looks like
         let placeholder = format!("__{:_<38}", name);
         let address = address.to_fixed_hex();
-        let matched = self.0.replace_all_in_place(&placeholder, &address);
-        if !matched {
+        if self.0.find(&placeholder).is_none() {
             return Err(LinkError::NotFound(name.to_string()));
         }
+        self.0 = self.0.replace(&placeholder, &address);
```
Comment (nlordell) [inline common/src/bytecode.rs:77]: Can we add a unit test similar to https://github.com/gnosis/ethcontract-rs/pull/162/files#diff-3da5111806a318876c8d54674237ec40L69 that checks that multiple occurrences of the same library get replaced?

**Review by nlordell — COMMENTED** [review]
@e00E, yeah lets move the `to_fixed_hash` method into `bytecode` module. Feel free to make it a function and not a extension trait method too.

File: common/src/bytecode.rs (hunk 1 above)
Comment (e00E) [reply common/src/bytecode.rs:77]: I added a test that makes sure that str::replace works as expected. This has less boilerplate than creating a whole ByteCode from scratch but let me know if that is what you had in mind.

File: common/src/bytecode.rs [hunk 2]
```diff
@@ -207,4 +210,34 @@ mod tests {
...
+            ),
+        ] {
+            let value: Address = value[2..].parse().unwrap();
+            assert_eq!(super::to_fixed_hex(&value), *expected);
+        }
+    }
+
+    #[test]
+    fn string_replacement_replaces_multiple_occurrences() {
+        // This is important for Bytecode::link.
```
Comment (nlordell) [inline common/src/bytecode.rs]: I would prefer if this method test `Bytecode::Link` directly. And maybe a second test to verify the branch where the library is not needed?

File: common/src/bytecode.rs [hunk 3]
```diff
@@ -71,11 +70,11 @@ impl Bytecode {
         //   '__$name__..__' with the library address; see generated bytecode for
         //   `LinkedContract` contract for and example of how it looks like
         let placeholder = format!("__{:_<38}", name);
-        let address = address.to_fixed_hex();
-        let matched = self.0.replace_all_in_place(&placeholder, &address);
-        if !matched {
+        let address = to_fixed_hex(&address);
+        if self.0.find(&placeholder).is_none() {
```
Comment (nlordell) [inline common/src/bytecode.rs]: nit: `!self.0.contains(&placeholder)`

**Review by nlordell — APPROVED** [review]
Looks great!

File: common/src/bytecode.rs [hunk 4]
```diff
@@ -207,4 +210,63 @@ mod tests {
...
+        let placeholder = format!("__{:_<38}", "name0");
+        let mut bytecode = Bytecode::from_hex_str(format!(
+            "0x61{}{}61{}",
+            placeholder, placeholder, placeholder
+        ))
+        .unwrap();
+        // name does not match
+        match bytecode.link("name1", address) {
+            Err(LinkError::NotFound(_)) => (),
+            _ => assert!(false, "should fail with not found error"),
```
Comment (nlordell) [inline common/src/bytecode.rs]: Its failing because of a clippy lint:
```suggestion
            _ => panic!("should fail with not found error"),
```

**Pull request merged**